src/info/repo.rs

```diff
@@ -16,159 +15,208 @@ use time::OffsetDateTime;
 use time_humanize::HumanTime;
 
 pub struct Repo<'a> {
-    repo: &'a Repository,
-    logs: Vec<Commit<'a>>,
+    git2_repo: &'a Repository,
+    repo: git::Repository,
+    authors: Vec<Author>,
+    total_num_authors: usize,
+    num_commits: usize,
+    time_of_most_recent_commit: git::actor::Time,
+    time_of_first_commit: git::actor::Time,
 }
 
-#[derive(Hash, PartialEq, Eq)]
 pub struct Sig {
-    name: String,
-    email: String,
+    name: git::bstr::BString,
+    email: git::bstr::BString,
+    email_lowercase: Option<String>,
 }
 
-impl From<Signature<'_>> for Sig {
-    fn from(sig: Signature) -> Self {
-        let name = String::from_utf8_lossy(sig.name_bytes()).into_owned();
-        let email = String::from_utf8_lossy(sig.email_bytes()).into_owned();
-        Self { name, email }
+impl From<git::actor::Signature> for Sig {
+    fn from(git::actor::Signature { name, email, .. }: git::actor::Signature) -> Self {
+        let needs_lowercase = email.chars().any(|c| c.to_ascii_lowercase() != c);
+        let email_lowercase = needs_lowercase.then(|| email.chars().collect::<String>());
+        Self {
+            name,
+            email,
+            email_lowercase,
+        }
     }
 }
 
-impl<'a> Repo<'a> {
-    pub fn new(
-        repo: &'a Repository,
-        no_merges: bool,
-        bot_regex_pattern: &Option<Regex>,
-    ) -> Result<Self> {
-        let logs = Self::get_logs(repo, no_merges, bot_regex_pattern)?;
-        Ok(Self { repo, logs })
+impl std::hash::Hash for Sig {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.email_lowercase
```

Comment:
IMO, the hash should be based on email_lowercase + name. Like that, we will get the same as `git shortlog -sne | wc -l` which is the de facto standard.

As is we are off by 6:

![number_of_authors](https://user-images.githubusercontent.com/13710835/161151698-3235713b-5343-4088-828e-cdccabeff032.png)

Reply:
Thanks for your guidance, fair enough! It's changed in d3d20eda4dd76720110f70b4992bb9c6483b1f12 .